src/smt.cpp

```diff
@@ -105,21 +112,199 @@ string or_omit(const expr &e) {
   return s;
 }
 
+ContextBuilder::ContextBuilder() {
+  use_z3 = false;
+}
 
-Expr Expr::urem(const Expr& rhs) const {
-  return {fmap(z3_expr, [&](auto e) { return z3::urem(e, *rhs.z3_expr); })};
+ContextBuilder& ContextBuilder::useZ3() {
+  use_z3 = true;
+  return *this;
 }
 
-Expr Expr::udiv(const Expr& rhs) const {
-  return {fmap(z3_expr, [&](auto e) { return z3::udiv(e, *rhs.z3_expr); })};
+std::optional<Context> ContextBuilder::build() const {
+  if (!use_z3) {
+    return std::nullopt;
+  }
+  return std::make_optional(Context(use_z3));
+}
+
+Context::Context() {
+  z3_ctx = nullptr;
+}
+
+Context::Context(bool use_z3) : Context() {
+  if (use_z3) {
+    z3_ctx = &ctx;
+  }
+}
+
+Expr Context::bvVal(const uint32_t val, const size_t sz) {
+  auto z3_expr = fmap(z3_ctx, [val, sz](auto ctx){ return ctx->bv_val(val, sz); });
+
+  return Expr(this, std::move(z3_expr));
+}
+
+Expr Context::bvConst(char* const name, const size_t sz) {
+  auto z3_expr = fmap(z3_ctx, [name, sz](auto ctx){ return ctx->bv_const(name, sz); });
+
+  return Expr(this, std::move(z3_expr));
+}
+
+Expr Context::boolVal(const bool val) {
+  auto z3_expr = fmap(z3_ctx, [val](auto ctx){ return ctx->bool_val(val); });
+
+  return Expr(this, std::move(z3_expr));
+}
+
+Expr::Expr(Context* const ctx, std::optional<z3::expr> &&z3_expr) : Expr(ctx) {
+  this->z3_expr = std::move(z3_expr);
+}
+
+Expr::Expr(Expr&& from) {
+  this->ctx = from.ctx;
+  this->z3_expr = std::move(from.z3_expr);
+}
+
+Expr Expr::clone() const {
+  auto cloned_z3_expr = this->z3_expr;
+  return Expr(this->ctx, std::move(cloned_z3_expr));
 }
 
 Expr Expr::simplify() const {
-  return {fmap(z3_expr, [](auto e) { return e.simplify(); })};
+  auto z3_expr = fmap(this->z3_expr, [](auto e) { return e.simplify(); });
+
+  return Expr(this->ctx, std::move(z3_expr));
 }
 
+ExprVec Expr::toNDIndices(const ExprVec &dims) const {
+  assert(dims.exprs.size() > 0);
+
+  auto idx_1d = this->clone();
+  auto expanded_exprs = ExprVec::withCapacity(this->ctx, dims.exprs.size());
+  std::for_each(dims.exprs.crbegin(), dims.exprs.crend(), 
```

Comment:
Changed from `std::accumulate` to `std::for_each`, as both `Expr` and `ExprVec` aren't very copy-friendly (`std::accumulate` [requires that the acc object is CopyConstructible](https://en.cppreference.com/w/cpp/algorithm/accumulate))